runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -67,40 +78,51 @@ public String getName() {
   @Override
   public void run(SourceContext<WindowedValue<T>> ctx) throws Exception {
     if (!(ctx instanceof StreamSource.ManualWatermarkContext)) {
-      throw new RuntimeException("We assume that all sources in Dataflow are EventTimeSourceFunction. " +
-          "Apparently " + this.name + " is not. Probably you should consider writing your own Wrapper for this source.");
+      throw new RuntimeException(
+          "We assume that all sources in Dataflow are EventTimeSourceFunction. " +
+              "Apparently " + this.name + " is not. " +
+              "Probably you should consider writing your own Wrapper for this source.");
     }
 
     context = (StreamSource.ManualWatermarkContext<WindowedValue<T>>) ctx;
     runtime = (StreamingRuntimeContext) getRuntimeContext();
 
-    this.isRunning = true;
+    isRunning = true;
+
+    reader = source.createReader(pipelineOptions, null);
```

Comment:
It seems like all we can do is to call `finalizeCheckpoint()` where appropriate. How does one snapshot the actual state? I think the interface needs to be extended to be more general.